### PR TITLE
Add pluggable controllers via ControllerProvider

### PR DESCRIPTION
**Pull request by ryannedolan**
Previously, a new adapter would require writing (or finding) an entirely new operator, or using one of the built-in reconcilers (e.g. KafkaTopicReconciler). This made it difficult to extend Hoptimator with new adapters or to customize the way the operator works.

Instead, we can make the operator pluggable, just as we have made the planner and catalog pluggable. A single adapter can now bundle catalog, planner, and operator plugins in the same place. This means we can drop in a custom Kafka adapter, for example, and get completely different logic for creating/deleting topics. (This turns out to be necessary for internal use-cases.)

An interesting side-effect of this is that the CLI now has access to control plane logic, which could be used to create/delete tables without running an operator.

### CODE REVIEW

File: hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/ControllerService.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.stream.Collectors;
+
+/** Loads controller plugins via the ControllerProvider Service Provider Interface. */
+public final class ControllerService {
+
+  public static Collection<ControllerProvider> providers() {
+    ServiceLoader<ControllerProvider> loader = ServiceLoader.load(ControllerProvider.class);
+    List<ControllerProvider> providers = new ArrayList<>();
+    loader.iterator().forEachRemaining(x -> providers.add(x));
+    return providers;
```
Comment (ehoner) [inline hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/ControllerService.java:20]: nit: 
```java
// loader variable isn't required
// change providers to lambda
List<ControllerProvider> providers = new ArrayList<>(); 
ServiceLoader.load(ControllerProvider.class).iterator().forEachRemaining(providers::add);
return providers;
```

FI alternative (requires JavaCompile > 8).
```java
return ServiceLoader.load(ControllerProvider.class)
    .stream()
    .map(ServiceLoader.Provider::get)
    .collect(Collectors.toList());
```

File: hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/Operator.java [hunk 2]
```diff
@@ -21,27 +21,36 @@
...
   private final Map<String, ApiInfo<?, ?>> apiInfo = new HashMap<>();
+  private final Properties properties;
 
-  public Operator(String namespace, ApiClient apiClient, SharedInformerFactory informerFactory) {
+  public Operator(String namespace, ApiClient apiClient, SharedInformerFactory informerFactory,
+      Properties properties) {
     this.namespace = namespace;
     this.apiClient = apiClient;
     this.informerFactory = informerFactory;
+    this.properties = properties;
```
Comment (ehoner) [inline hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/Operator.java:44]: I think it is better to avoid potentially `null` fields. Add test for input or require non-null. ie. 
```
# test
this.properties = properties == null ? new Properties() : properties;

# validate
this.properties = Validate.notNull(properties);
```

File: hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/HoptimatorOperatorApp.java [hunk 3]
```diff
@@ -6,29 +6,42 @@
...
+    this.properties = properties;
+  }
+
   public static void main(String[] args) throws Exception {
     if (args.length != 1) {
       throw new IllegalArgumentException("Missing model file argument.");
     }
-    HoptimatorPlanner.Factory plannerFactory = HoptimatorPlanner.Factory.fromModelFile(args[0],
-        new Properties());
+    new HoptimatorOperatorApp(args[0], "default", new Properties()).run();
```
Comment (ehoner) [inline hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/HoptimatorOperatorApp.java:39]: Is it useful to allow overrides from the command line? If `new Properties()` is replaced with `System.getProperties()`, then `-Dhoptimator.key=value` settings would be honored with `run.sh`. Execution would be similar to:
```sh
$ JAVA_OPTS=-Dhoptimator.cli=value hoptimator-operator/run.sh <model>
```

This might help address the `namespace` TODO.

**Review by ehoner — CHANGES_REQUESTED** [review]
The nits are optional. The other comments can be discussed or fixed.

File: hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/HoptimatorOperatorApp.java (hunk 3 above)
Comment (ryannedolan) [reply hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/HoptimatorOperatorApp.java:39]: I'd rather keep a distinction between JAVA_OPTS and app options. I opened issue #17 to add `--properties key=value` or similar at some point.

File: hoptimator-cli/src/main/resources/log4j.properties [hunk 4]
```diff
@@ -0,0 +1,7 @@
+log4j.rootLogger=ERROR,stderr
+
+log4j.appender.stderr=org.apache.log4j.ConsoleAppender
+log4j.appender.stderr.layout=org.apache.log4j.PatternLayout
+log4j.appender.stderr.layout.ConversionPattern=%p\t%C{1}\t%m%n
```
[the remark this answers is not in the document]
Comment (ryannedolan) [reply hoptimator-cli/src/main/resources/log4j.properties:5]: In the CLI it's probably just noise. I can add timestamps to the operator logs tho.

File: hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/ControllerService.java (hunk 1 above)
Comment (ryannedolan) [reply hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/ControllerService.java:20]: Looks like this depends on java 9, which is probably okay, but let's save that for a future PR. We can clean up a lot of small stuff like this if we skip to 11.

File: hoptimator-cli/src/main/resources/log4j.properties (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryannedolan) [reply hoptimator-cli/src/main/resources/log4j.properties:5]: I dropped log4j in favor of slf4j-simple, so deleting this file.

**Review by ehoner — APPROVED** [review]
LGTM
